README.md
ITK Cooling Scheme
========================
This is an analytical model of the thermal model, based off of the analytical barrel model from
Georg Viehhauser.

More information is available in the twiki https://wiki-zeuthen.desy.de/ATLAS/ThermalModelForPetal

Checking Out and Running
--------
To checkout, do:

    git clone ssh://git@gitlab.cern.ch:7999/desy-atlas/itk/StripsThermalModel.git

To run, do :

    cd macros
    python RunModel.py --cooling -35

Submodules
--------
The following submodules are defined in the python directory (modules with a * are fairly large) :

Submodules                   | Purpose                                          | Assignee   | Status
-----------------------------|--------------------------------------------------|------------|------------
0. GlobalSettings            | Settings for the analysis: nyears, nstep, step.. | Kurt       | Done
1. SafetyFactors             | for layout, fluence, thermal impedance, I, vbias | Kurt       | Done
2. Layout                    | Number of staves per end                         | Claire     | Done
3. PoweringEfficiency        | FEAST efficiency, DCDC2 efficiency               | Kurt *     | Done
4. CableLosses               | Service modules and outer cables                 | Claire     | Done
(4.) FrontEndComponents      | Front-end components, before irradiation         | (Claire)   | Done
5. EOSComponents             | lpbt, gbtia, gbld (before irradiation)           | Kurt       | Done
6. AbcTidBump                | Increase in digital current in ABC               | Claire *   | Done
7. SensorProperties          | Sensor area, bias volate, resistors              | Kurt       | Done
8. NominalPower              | Expressions for the power in each element        | Claire *   | Done
9. ThermalImpedances         | Fit for impedances (R) given simulated points    | Kurt *     | Done
10. Temperatures             | Activation, barrel LS and SS                     | Claire     | Done
11. OperationalProfiles      | Luminosity (inst, int), efficiency (?)           | Kurt *     | Done
12. SensorLeakage            | Leakage current as a function of flux (A/cm^2)   | Claire *   | Done
13. SensorTemperatureCalc    | Sensor temperature calculation                   | (Both)     | Done, debugged w/B1

Plotting/Checks/Investigations
-----------
Topic                        | Purpose                                          | Assignee   | Status
-----------------------------|--------------------------------------------------|------------|------------
Sensor power                 | Plotting to check correct implementation         | Claire     | Done plotting
TID bump                     | Plot to check and investigate the modelling      | Claire     | Done plotting
Feast efficiency             | Plot to check and investigate the modelling      | Kurt       | Done
Operating profiles           | Reproducing plots from Graham and Georg          | Kurt       | Done
SensorTemperatureCalc tsensor    | Sensor temperature                           | Kurt       | Done
SensorTemperatureCalc tabc       | ABC temperature                              | Kurt       | Done
SensorTemperatureCalc thcc       | HCC temperature                              | Kurt       | Done
SensorTemperatureCalc tfeast     | FEAST temperature                            | Kurt       | Done
SensorTemperatureCalc teos       | EOS temperature                              | Kurt       | Done
SensorTemperatureCalc pmodule    | Power per module (front-end + HV)            | Kurt       | Done
SensorTemperatureCalc pmtape     | Power loss in tape per module                | Kurt       | Done
SensorTemperatureCalc pmhv       | HV power per module (leakage + resistors)    | Kurt       | Done
SensorTemperatureCalc isensor    | Sensor current                               | Kurt       | Done
SensorTemperatureCalc pmhvr      | HV power per module due to serial resistors  | Kurt       | Done
SensorTemperatureCalc powertotal | Total power in layer                         | Kurt       | Done
SensorTemperatureCalc phvtotal   | Total HV Power (sensor+resistors) in layer   | Kurt       | Done
SensorTemperatureCalc pmhvmux    | HV Power parallel resistor                   | Kurt       | Done
SensorTemperatureCalc itape      | Tape current per module                      | Kurt       | Done
SensorTemperatureCalc idig       | Digital current per module                   | Kurt       | Done
SensorTemperatureCalc efffeast   | FEAST efficiency                             | Kurt       | Done
SensorTemperatureCalc ptape      | Power loss in complete tape in layer         | Kurt       | Done
SensorTemperatureCalc pstave     | Stave Power in layer                         | Kurt       | Done

More plots -- plotting and debugging (focus on flat-35 scenario)
-----------
Topic                                 | Notes                                                                              | Assignee   | Status
--------------------------------------|------------------------------------------------------------------------------------|------------|------------
Sensor temperature                    | p. 11 in https://indico.cern.ch/event/625365                                       | Kurt       | Done, debugged
Power summary plot                    | p. 11 of slides; total power, w/o HV, w/o tape loss                                | Kurt       | Done, debugged
Total HV power per module             | p. 11 of slides                                                                    | Kurt       | Done, debugged
Temperature summary of all components | p. 12 of slides                                                                    | Kurt       | Done, debugged
Total power in B1                     | p. 12 of slides                                                                    | Kurt       | Done, debugged

Simple numbers that need to be switched from barrel to endcap values:
-----
 - **Fluences/Total ionizing dose for the petal**. Later, extend this to include dependence on eta and
radius (or equivalent parameterization).
 - **Layout**: change the number of staves with the different petal sensors..

Data / fits that we need
-----
 - **FEAST efficiency** (vs temperature, current): Georg et al have an initial fit based on some data,
but the fit could be improved and more precise data could be pushed for.
 - **Thermal impedances**: These are currently based on thermal simulations from the the barrel, and
need to be rederived using endcap thermal simulations.
 - **Losses**: no value for losses in tapes? 

Questions
-----
 - **Sensor leakage**: line 28 
 - Document (v2.1): Eq. 15a has I instead of I^2 ?
 - **Operational profiles**: Figures 11 and 12 of report v2.3 are not matching associated plots in Mathematica file
(Section 11) E.g. second plateau Year 8 of B1 is around 15000 kRad in the document, but 20000 kRad in Mathematica's plot.
Was there an update in between? 


Other info
-----
 - Data files are kept in the "data" directory -- an example of the format is FeastEfficiencyData.txt.


Mathematica-to-python Quick-guide
-----
 - "./x -> y" means replace all "x" with "y"
 - "x[y_] := 2y": set delay, evaluated at call. Python analog: def x(y) : return 2y
 - "x = 2y": set, meaning x is set immediately to 2y, whatever y is at the time. Python analog: x = 2y
 - Log in mathematica is natural logarithm (base e)


ChangeLog
2017-09-24 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Fix barrel plots
   * Fix module-by-module barrel result tables
   * Fixing module-by-module barrel treatment for config printing

2017-09-22 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * First attempt to run full barrel on a module-by-module basis (some bugs remain)

2017-09-19 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Rewrite TID module to accommodate multiple TID parameterizations
   * Improve Feast efficiency plots
   * Improve RunModel.py

2017-09-18 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Add first iteration of new TID parameterization from Kyle (not the default)
   * A plot to compare the new and old parameterization

2017-09-17 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Add PoweringEfficiency.ModelVersion configuration (v00,v01)
   * Save GitHash information to configuration list
   * Change default HVMUX to 10M ohms (from 1M)
   * Improve CompareScenarios.py
   * Add Vfeast and R3 config files as default (these have been used since ITK in June)

2017-09-15 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Automatic hlines... also propagate thermal runaway info vs module
   * Propagate thermal runaway all the way through to output tables
   * Add a script to put together scenario tables
   * Add Results.npy and Config.npy for saving results (needed for new scenario script)

2017-09-14 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Include cable losses in total power estimates (mistakenly omitted)
   * Split safetycurrent into digital (safetycurrentd, should be 20%) and analog (safetycurrenta, should be 5%)
   * Remove vbiasscale in favor of just setting vbias directly (cleans things up a bit)
   * Remove safetylayout since we don't use it
   * Current safety factor was double-counted for Pamac (small effect)

2017-09-12 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Clarify some additional quatities (e.g. EOS not included in some power numbers)
   * Add totals (per petal, full system) to table printouts
   * More compact endcap configuration printed to latex
   * Add current load on tape due to EOS

2017-09-09 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Add petal power plot

2017-09-08 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Fix HV quantities (one small bug in a power summary plot) and try to clarify the labeling
   * Clean up code for printing out latex tables.

2017-08-30 Kurt Brendlinger <kurt.brendlinger@cern.ch>
   * Remove incorrect quantities from SensorTemperatureCalc: powertotal, pstave, phvtotal
   * Fix up itape, ptape
   * New method to transmit the tape load of the previous modules (e.g. R0, R1) to the next module R2
   * Add itape_cumulative, ptape_cumulative which include the load of the previous modules
   * Add TotalPower into ExtendedModelSummaryPlots
   * Add TotalHVPower into ExtendedModelSummaryPlots

2017-08-18 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Un-hardcode "year 14" as last year
    * Add first latex document output
    * Remove LV current output (it is the same as itape)
    * Print out summary quantities at start, TID, end-of-life
    * Add Itape_eos, tape current load due to EOS
    * Add Veos (EOS input voltage) -- 11V
    * Add units to Vfeast

2017-07-23 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Shuffle some definitions in NominalPower.py
    * More options for printing out plots in ExtendedRunModel.py
    * Add LV current output
    * Clarify some of the quantities

2017-07-21 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Print out more configuration details

2017-07-07 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Print out more configuration details
    * Description of variables

2017-07-04 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Clean up methods for output directory

2017-07-03 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Add output tables for endcap ExtendRunModel
    * Some other small changes
    * Consolidate "configuration by command-line" treatment
    * Add SafetyFactor command-line

2017-07-02 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Make most values configurable
    * Add a table printout (for latex documents)

2017-06-28 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Add TID characterization plots vs time

2017-06-26 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Fix up some plots; add propaganda plots
    * Add operational profile plots for endcaps

2017-06-25 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Add new TID and Flux parameterization
    * (See https://twiki.cern.ch/twiki/bin/view/Atlas/RadiationBackgroundSimulationsFullyInclinedAt4)

2017-06-23 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Fix up plot ranges
    * Clean up SensorTemperatureCalc.py

2017-06-22 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Add Qsensor and TCoolant headroom plots
    * Fix bug in SensorTemperatureCalc related to solving for Qsensor (Qref)
    * Add thermal balance illustration plots

2017-06-14 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Fix bug in FEAST efficiency plot, affecting the 2-FEAST R3 scenario.

2017-06-13 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Add capability to model a module with 2 feasts for the ABCs/HCCs instead of 1
    * Un-fix bug in parameterization of TID (2017-06-11) -- values were ok after all.
    * Add CompareScenarios.py macro
    * Make Safety Factors configurable
    * Fix issue in reloading python modules (order matters)
    * Add thermal impedances calculated from Yu-Heng's numbers
    * Add updated thermal impedance plots

2017-06-12 Kurt Brendlinger <kurt.brendlinger@cern.ch>
    * Write config file to output directory so you know what configuration was used to make the plots
    * Fix a certain failure mode in thermal runaway scenario (SensorTemperatureCalc)
    * AddRunParameterLabels function for plots, (like safety factor info)

2017-06-11 Kurt Brendlinger <kurt.brendlinger@cern.ch>

python/PlotUtils.py
#!/usr/bin/env python

import ROOT
import os,sys

# To later defined gStyle common stuff for all plots

# Does not seem to work...
def ApplyGlobalStyle() :
    mystyle = ROOT.TStyle("mystyle","mystyle")
    mystyle.SetMarkerStyle(20)
    mystyle.SetMarkerSize(1)
    mystyle.SetCanvasColor(0)
    mystyle.SetPadBorderMode(0)
    mystyle.SetCanvasBorderMode(0)
    mystyle.SetFrameBorderMode(0)
    mystyle.SetOptTitle(0)

    mystyle.SetPadTopMargin(0.05)
    mystyle.SetPadRightMargin(0.05)
    mystyle.SetPadBottomMargin(0.11)
    mystyle.SetPadLeftMargin(0.16)

    # all axes
    mystyle.SetTitleSize  (22   ,'xyz')
    mystyle.SetTitleFont  (43   ,'xyz')
    mystyle.SetLabelSize  (22   ,'xyz')
    mystyle.SetLabelFont  (43   ,'xyz')

    # x axis
    mystyle.SetTitleXOffset(1.0)
    mystyle.SetLabelOffset(0.003,'x')
    # y axis
    mystyle.SetTitleOffset(1.75 ,'y')
    mystyle.SetLabelOffset(0.003,'y')

    # line styles for the extended model
    mystyle.SetLineStyleString(10,'60 10');
    mystyle.SetLineStyleString(11,'50 10');
    mystyle.SetLineStyleString(12,'40 10');
    mystyle.SetLineStyleString(13,'30 10');
    mystyle.SetLineStyleString(14,'20 10');

    mystyle.SetLineStyleString(15,'30 10');
    mystyle.SetLineStyleString(16,'20 10');
    mystyle.SetLineStyleString(17,'10 10');
    mystyle.SetLineStyleString(18, '5 10');

    # For intermediate modules
    mystyle.SetLineStyleString(19, '5 5');

    ROOT.gROOT.SetStyle("mystyle")

def SetStyleTitles(tobject, title, xtitle, ytitle) :
    
    print "Setting titles for plot: %s" % title
    
    tobject.SetTitle(title)
    tobject.GetXaxis().SetTitle(xtitle)
    tobject.GetXaxis().SetTitleOffset(1.3)
    tobject.GetYaxis().SetTitle(ytitle)
    tobject.GetYaxis().SetTitleOffset(1.3)
    
# 2D objects (with z axis)
def Set2DStyleTitles(tobject, title, xtitle, ytitle, ztitle) :
    
    print "Setting titles for plot: %s" % title
    
    tobject.SetTitle(title)
    tobject.GetXaxis().SetTitle(xtitle)
    tobject.GetXaxis().SetTitleOffset(1.7)
    tobject.GetYaxis().SetTitle(ytitle)
    tobject.GetYaxis().SetTitleOffset(1.7)
    tobject.GetZaxis().SetTitle(ztitle)
    tobject.GetZaxis().SetTitleOffset(1.4)
    
    
def SetStyleLegend(leg) :
    
    leg.SetName('legend')
    leg.SetBorderSize(0)
    leg.SetFillStyle(0)
    leg.SetTextSize(0.035)
    leg.SetTextFont(42)
    
def MakeGraph(name,title,xtitle,ytitle,xlist,ylist,xerrlist=None,yerrlist=None) :
    from array import array

    if (len(xlist) != len(ylist)) :
        print 'Error! List sizes are not correct for graph %s (%d vs %d)'%(name,len(xlist),len(ylist))
        import sys; sys.exit()
    if not xerrlist and not yerrlist :
        graph = ROOT.TGraph(len(xlist),array('d',xlist),array('d',ylist))
    else :
        graph = ROOT.TGraphErrors(len(xlist),array('d',xlist),array('d',ylist),array('d',xerrlist),array('d',yerrlist))
    graph.SetName(name)
    graph.SetTitle(title)
    graph.GetXaxis().SetTitle(xtitle)
    graph.GetYaxis().SetTitle(ytitle)
    graph.SetLineWidth(3)
    return graph

def GetCoolingOutputTag(cooling_option) :
    if 'flat' in cooling_option :
        return cooling_option.replace('flat','flat_').replace('-','m')
    if 'ramp' in cooling_option :
        return cooling_option.replace('ramp','ramp_').replace('-','m')
    return 'unknownScenario'

def GetCoolingScenarioLabel(cooling_option) :
    if 'flat' in cooling_option :
        temp = int(cooling_option.replace('flat',''))
        return ('Flat %s#circ cooling scenario'%(temp)).replace('-','#minus')
    if 'ramp' in cooling_option :
        temp = int(cooling_option.replace('ramp',''))
        return ('Ramp %s#circ cooling scenario'%(temp)).replace('-','#minus')
    return 'unknown cooling scenario'

def AddRunParameterLabels(legend,additionalinfo=[],wrap=False) :
    import SafetyFactors
    import CoolantTemperature
    layout,fluence,thermalimpedance,currenta,currentd,vbias = '','','','','',''

    if SafetyFactors.safetyfluence :
        fluence = 'flux_{ }#times^{ }%2.1f;  '%(SafetyFactors.safetyfluence + 1.0)
    if SafetyFactors.safetythermalimpedance :
        thermalimpedance = '^{}R_{thermal }#times^{ }%2.1f;  '%(SafetyFactors.safetythermalimpedance + 1.0)
    if SafetyFactors.safetycurrentd :
        currentd = 'I_{dig}#times%2.2f;  '%(SafetyFactors.safetycurrentd + 1.0)
    if SafetyFactors.safetycurrenta :
        currenta = '^{}I_{a }#times^{ }%2.2f;  '%(SafetyFactors.safetycurrenta + 1.0)
    if (SafetyFactors.vbias) != 500. :
        vbias = '^{}V_{bias}=^{ }%2.0f;  '%(SafetyFactors.vbias)

    #text = ('%s%s%s%s%s'%(layout,fluence,thermalimpedance,current,vbias)).rstrip('; ')
    sf1 = ('%s%s%s'    %(layout,fluence,thermalimpedance)).rstrip('; ')
    sf2 = ('%s%s%s'    %(                               currentd,currenta,vbias)).rstrip('; ')
    no_factors = ' none' if (not sf1 and not sf2) else ''

    nlines = 0
    if True :
        legend.AddEntry(0,GetCoolingScenarioLabel(CoolantTemperature.cooling),'')
        nlines += 1
    if True :
        legend.AddEntry(0,'Safety factors:%s'%(no_factors),'')
        nlines += 1
    if sf1 :
        legend.AddEntry(0,sf1,'')
        nlines += 1
    if sf2 :
        legend.AddEntry(0,sf2,'')
        nlines += 1
    for addinfo in additionalinfo :
        if not addinfo :
            continue

        add_label = [addinfo,'']
        if wrap :
            while(len(add_label[0]) > 28) :
                tmp = add_label[0].split(' ')
                add_label[0] = ' '.join(tmp[:-1])
                add_label[1] = ' '.join([tmp[-1],add_label[1]])

        legend.AddEntry(0,add_label[0],'')
        nlines += 1
        if add_label[1] :
            legend.AddEntry(0,add_label[1],'')
            nlines += 1

    while nlines < 5 :
        legend.AddEntry(0,'','')
        nlines += 1

    return

def ColorPalette() :
    from ROOT import kBlack,kRed,kBlue,kAzure,kGreen,kMagenta,kCyan,kOrange,kGray,kYellow
    return [kBlack+0,kRed+1,kAzure-2,kGreen-6,kMagenta-6,kCyan+1,kOrange+1
            ,kGray+1,kRed-6,kBlue+3,kGreen+3,kMagenta+3,kCyan+3,kOrange+3
            ,kGray,kRed-7,kBlue-7,kGreen-7,kMagenta-7,kCyan-7,kOrange-7
            ,kYellow+2,kRed-5,kBlue-5,kGreen-5,kMagenta-5,kCyan-5,kOrange-5
            ,21,22,23,24,25,26,27,28,29,30
            ,21,22,23,24,25,26,27,28,29,30
            ,21,22,23,24,25,26,27,28,29,30
            ]

def ColorGradient(i,ntotal) :
    import ROOT
    if ntotal == 1 :
        return ROOT.kBlack
    from array import array
    NCont = 255
    # Blue -> Green -> Red
#     NRGBs = 3
#     stops = array('d',[ 0.00, 0.50, 1.00 ])
#     red   = array('d',[ 0.00, 0.50, 1.00 ])
#     green = array('d',[ 0.00, 1.00, 0.00 ])
#     blue  = array('d',[ 1.00, 0.50, 0.00 ])
#     ROOT.TColor.CreateGradientColorTable(NRGBs, stops, red, green, blue, NCont)

    # Some weird default
#     NRGBs = 5
#     stops = array('d',[  0.00, 0.34, 0.61, 0.84, 1.00 ])
#     red   = array('d',[  0.00, 0.00, 0.87, 1.00, 0.51 ])
#     green = array('d',[  0.00, 0.81, 1.00, 0.20, 0.00 ])
#     blue  = array('d',[  0.51, 1.00, 0.12, 0.00, 0.00 ])
#     ROOT.TColor.CreateGradientColorTable(NRGBs, stops, red, green, blue, NCont)

    # Green -> Red
#     NRGBs = 2
#     stops = array('d',[ 0.00,1.00 ])
#     red   = array('d',[ 0.50,1.00 ])
#     green = array('d',[ 1.00,0.00 ])
#     blue  = array('d',[ 0.50,0.00 ])
#     ROOT.TColor.CreateGradientColorTable(NRGBs, stops, red, green, blue, NCont)

    # Blue -> Red (but it doesn't go through white)
    NRGBs = 2
    stops = array('d',[ 0.00,1.00 ])
    red   = array('d',[ 0.10,0.90 ])
    green = array('d',[ 0.10,0.10 ])
    blue  = array('d',[ 1.00,0.10 ])
    ROOT.TColor.CreateGradientColorTable(NRGBs, stops, red, green, blue, NCont)

    #ROOT.gStyle.SetPalette(ROOT.kThermometer)
    ROOT.gStyle.SetNumberContours(255)
    the_int = int(254*i/float(ntotal-1))
    return ROOT.gStyle.GetColorPalette(the_int)

def AddToStack(stack,leg,h) :
    h.SetFillColor(ColorPalette()[stack.GetNhists()+1])
    h.SetLineWidth(2)
    h.SetDrawOption('l')
    stack.Add(h)
    #leg.AddEntry(h,h.GetTitle().replace('power ','').replace('power',''),'f')
    return h

def MakePlotMinimumZero(plotname) :
    return plotname in ['tc_headroom','idig','ifeast','ifeast_in',
                        'phv_wleakage','phvr','phvmux','hv_power_resistors','pmodule_noHV',
                        'pmodule','powertotal','itape','itape_cumulative','itape_eos','ptape','ptape_cumulative','pstave',
                        'phvtotal','ppetal']

def GetPlotForcedMinimum(plotname) :
    return {'qsensor_headroom': 0.09}.get(plotname,None)

def GetOutputPath(modulename,options) :
    just_outside_git_repository = os.getcwd().split('/StripsThermalModel')[0]

    # Move output path outside code directory
    outputpath = '%s/plots/%s'%(just_outside_git_repository,modulename)

    # If a different output name is specified
    if hasattr(options,'outdir') :
        outputpath = '%s/plots/%s'%(just_outside_git_repository,options.outdir)

    if not os.path.exists(outputpath) :
        os.makedirs(outputpath)

    # print '%s output written to %s'%(modulename,outputpath)
    return outputpath

# A quick, useful helper function
def GetResultDictIndex(names,ring_mod,disk_layer) :
    try :
        index = names.index('R%dD%d'%(ring_mod,disk_layer))
    except ValueError :
        index = names.index('L%dM%d'%(disk_layer,ring_mod))
    return index

# A quick, useful helper function
def GetResultDictIndexInverted(names,ring_lay,disk_mod,isEndcap) :
    ring_mod   = ring_lay if isEndcap else disk_mod
    disk_layer = disk_mod if isEndcap else ring_lay
    try :
        index = names.index('R%dD%d'%(ring_mod,disk_layer))
    except ValueError :
        index = names.index('L%dM%d'%(disk_layer,ring_mod))
    return index
